MAINT: backports/prep for 1.13.1

tylerjereddy opened the pull request:
We're about 3.5 weeks from the planned branching for SciPy `1.14.x`. Since `1.13.0` was a bit rushed to support NumPy 2.x, and indeed I made a few oversights for backporting, it may make sense to at least assess how much work it would be to deliver `1.13.1`. I've cleaned up the backport labels a bit too, since there were almost 50 of them. Many were things I could have backported to `1.12.x`, but I think that would stretch me (and reviewers/support) too thin at this point.

Backports included here (so far):

1. gh-20322
2. gh-20333
3. gh-20401
4. gh-20435
5. gh-20437
6. gh-20449
7. gh-20473
8. gh-20474
9. gh-20484
10. gh-20505 (already covered by gh-20537)
11. gh-20516
12. gh-20527 (**danger: I applied the patch manually to the old version of the `special` code**; I did backport the new test though, and it did fail before/pass after my manual application of a slightly different patch)
13. gh-20530
14. gh-20569
15. gh-20573
16. gh-20607
17. gh-20611
18. gh-20633
19. gh-20653
20. gh-20654
21. gh-20444
22. gh-20687
23. gh-20280
24. gh-20601
25. gh-20629
26. gh-20644
27. gh-20727
28. gh-20533
29. gh-20592
30. gh-20702

Backports already merged to the release branch:
1. gh-20567
2. gh-20537

TODO:

- [x] perhaps deal with what I believe is NumPy copy semantics changes causing a few fails with pre-release (https://github.com/scipy/scipy/actions/runs/9102530687/job/25022358623?pr=20632), though I believe we don't really have to do that on the release branch and could also pin NumPy back a bit (maybe I'll evaluate size of patch needed locally...)
- [x] CI failure from https://github.com/scipy/scipy/issues/20714 (persists in `main` at time of writing)
- [x] https://github.com/scipy/scipy/issues/20576 (at least one of those is showing up here already I think)
- [x] there are a few 1.13.0 performance/hang-related tickets open to keep an eye on I think (OpenBLAS and/or Qhull thread contention?)
- [x] gradually running more of the CI here, including wheel build flush eventually because of i.e., OpenBLAS bump
- [x] check for more backport candidates from the last few days
- [x] update the release notes again
- [x] read the diff carefully
- [x] (minor) `doc/source/.jupyterlite.doit.db` isn't gitignored on this branch/PR when I do the doc build (which otherwise passes at the time of writing) locally
- [x] backport linter suppressions for stuff like `UP031 Use format specifiers ...` maybe?

Commit:
MAINT: added doc/source/.jupyterlite.doit.db to .gitignore See #20264

tylerjereddy committed:
BUG: sync pocketfft again

* Fixes gh-20300 by syncing `pocketfft` again, this time
to completely disable `aligned_alloc`; see https://github.com/scipy/pocketfft/pull/3
for details, but in short our more conservative shim
was not sufficient for `conda-forge`, so let's just do the same
thing NumPy did...

[skip cirrus] [skip circle]

AtsushiSakai committed:
DOC: optimize: fix wrong optional argument name in `root(method="lm")`. (#20401)

j-bowhay committed:
DOC: add missing deprecations from 1.13.0 release notes

[doc only]

lucascolley committed:
MAINT/DOC: fix syntax in 1.13.0 release notes (#20437)

j-bowhay committed:
DOC: remove spurious backtic from release notes

[doc only]

ev-br committed:
BUG: linalg: fix ordering of complex conj gen eigenvalues

This was reported to cause test failures under windows + MKL in conda-forge
cf https://github.com/scipy/scipy/issues/20471

ev-br committed:
TST: interpolate: bump test tolerance in TestRGI::test_derivatives

A slight tolerance violation was reported on conda-forge in
https://github.com/scipy/scipy/issues/20472

ev-br committed:
TST: interpolate: bump test tolerance in TestInterpN

A small tolerance violation reported on conda-forge
in https://github.com/scipy/scipy/issues/20472

ev-br committed:
TST: interpolate: bump test tolerance in test_rbfinterp::test_pickleable

An exact equality was reported as flaky on conda-forge in
https://github.com/scipy/scipy/issues/20472

The tolerance violations are of the order of fp noise (< 2e-16), and I don't
think that pickling/unpickling guarantees bit-to-bit compatibility.
In principle, this may invoke recalculations and those may be subject
to fp noise. So I think it's OK to only require allclose(atol=eps)
instead of exact equality.

thalassemia committed:
TST: compare absolute values of U and VT in pydata-sparse SVD test

dschult committed:
align dok.pop() to dict.pop() case with no default (#20322)

ev-br committed:
TST: sparse/linalg: bump tolerance for an svds test

tol violation observed on conda-forge on win+blis; suggested in
https://github.com/scipy/scipy/pull/20474#issuecomment-2054269010

ev-br committed:
BUG: linalg: fix eigh(1x1 array, driver='evd') f2py check (#20516)

f2py check was just too strict;
LAPACK docs indicate that for N=1, lwork>=1 is acceptable:

*  LWORK   (input) INTEGER
*          The dimension of the array WORK.
*          If N <= 1,               LWORK must be at least 1.
*          If JOBZ = 'N' and N > 1, LWORK must be at least 2*N+1.
*          If JOBZ = 'V' and N > 1, LWORK must be at least
*                                                1 + 6*N + 2*N**2.

https://www.netlib.org/lapack/explore-3.1.1-html/ssyevd.f.html

closes gh-20512

bnavigator committed:
BLD: Fix error message for f2py generation fail (#20530)

Co-authored-by: Ralf Gommers <ralf.gommers@gmail.com>

PicoCentauri committed:
BUG: fix spherical_in for n=1 and z=0 (#20527)

thalassemia committed:
TST: Adapt to `__array__(copy=True)`

[skip circle]

mattip committed:
update openblas to 0.3.27

scottshambaugh committed:
BUG: Fix error with 180 degree rotation in Rotation.align_vectors() with an infinite weight (#20573)

* Fix exact rotations at 180 deg, improve near 180 deg

Comments

* Tests for exact near 180 deg rotations

* Fix tests

* Code review updates

---------

Co-authored-by: Scott Shambaugh <scottshambaugh@users.noreply.github.com>

lucascolley committed:
MAINT: lint: temporarily disable UP031
[lint only]

h-vetinari committed:
BUG: special: handle uint arrays in factorial{,2,k} (#20607)

rgommers committed:
DEV: lint: disable UP031

Showed up as a linting error in an unrelated PR for me:
```
scipy/interpolate/_interpolate.py:918:30: UP032 [*] Use f-string instead of `format` call
scipy/interpolate/_interpolate.py:1972:30: UP032 [*] Use f-string instead of `format` call
```
This should not happen; the old code is fine, so this check needs to be
silenced or fixed separately. Similar to gh-20601.

[skip ci]

tylerjereddy committed:
BUG: spherical_in old patch

* Apply "old version" of the patch provided
at gh-20527

tylerjereddy committed:
DOC: update 1.13.1 relnotes

* Update the SciPy `1.13.1` release notes following
a series of backports.

[skip cirrus]

tylerjereddy committed:
BUG: fix Vor/Delaunay segfaults

* Deals with the `spatial` part of gh-20623 (`Voronoi`
was also affected beyond the originally-reported `Delaunay`
segfault).

* Both classes are documented to accept arrays with two dimensions
only, so raise a `ValueError` in cases with other dimensions to
avoid the segfault.

* Other potential points of confusion here are the differences
between arrays with two dimensions and two dimensional arrays
that represent generators with more than two dimensions via
the number of columns, but this is just how things are for
most array programming languages of course. Also, the docs
don't explicitly say array-like for the generators I don't think, but this patch
only worked if I placed it after the coercion to array type.

[skip circle] [skip ci] [ci skip]

andfoy committed:
BUG: prevent QHull message stream being closed twice (#20611)

rgommers commented:
This all LGTM so far. 

> there are a few 1.13.0 performance/hang-related tickets open to keep an eye on I think (OpenBLAS and/or Qhull thread contention?)

OpenBLAS is fine already I think, given the version is bumped to 0.3.27 in this PR. For Qhull, it'd be useful to backport gh-20611 since that makes the full test suite pass with free-threaded CPython and is a small/safe fix. gh-20619 could be backported once merged, but it's noncritical for 1.13.1

tylerjereddy committed:
BUG: value_indices unfixed types

* Fixes gh-19423

* Add a few more `case` statements to account for
the (i.e., Windows) data types that don't have a fixed
width, and add a regression test.

[skip circle] [skip cirrus]

tylerjereddy commented:
I added the backport Ralf suggested and the other `spatial` segfault fix from the last few days, so we're back at parity with the last of merged backport PRs.

That said, I'm skipping the CI for now because it is known to fail per the TODO list in the original comment above. I'll try to chip away at that as time permits.

grlee77 committed:
fix origin handling for minimum_filter and maximum_filter with axes subset

mdhaber committed:
MAINT: stats.yulesimon: fix kurtosis

dschmitz89 committed:
MAINT: stats: update boost to improve `skewnorm.ppf` (#20643)

* MAINT: stats: update boost to fix improve `skewnorm.ppf`

dschult committed:
BUG: sparse: Clean up 1D input handling to sparse array/matrix and add tests (#20444)

* test 1d input to sparse. Add FutureWarnings and ValueErrors

* remove matrix changes. Let them create 2D

* correct imports

* rebase on main and update to support 1D CSR input

tylerjereddy commented:
@steppi @czgdp1807 @dschmitz89 what do you think about applying the patch below (similar to the one suggested at: https://github.com/scipy/scipy/issues/20208#issuecomment-2027547367) to resolve the first part of https://github.com/scipy/scipy/issues/20576 on the release branch here? There's an explanation there about why I'm hesitant to backport gh-20393 -- do you agree with my hesitation?

<details>

```diff
--- a/scipy/stats/_continuous_distns.py
+++ b/scipy/stats/_continuous_distns.py
@@ -692,12 +692,10 @@ class beta_gen(rv_continuous):
         return _boost._beta_sf(x, a, b)
 
     def _isf(self, x, a, b):
-        with np.errstate(over='ignore'):  # see gh-17432
-            return _boost._beta_isf(x, a, b)
+        return sc.betaincinv(a, b, 1 - x)
 
     def _ppf(self, q, a, b):
-        with np.errstate(over='ignore'):  # see gh-17432
-            return _boost._beta_ppf(q, a, b)
+        return sc.betaincinv(a, b, q)
 
     def _stats(self, a, b):
         return (
diff --git a/scipy/stats/tests/test_distributions.py b/scipy/stats/tests/test_distributions.py
index 6d91bb8a6..de26dc3d8 100644
--- a/scipy/stats/tests/test_distributions.py
+++ b/scipy/stats/tests/test_distributions.py
@@ -4454,7 +4454,7 @@ class TestBeta:
         assert_equal(stats.beta.pdf(1, a, b), 5)
         assert_equal(stats.beta.pdf(1-1e-310, a, b), 5)
 
-    @pytest.mark.xfail(IS_PYPY, reason="Does not convert boost warning")
+    @pytest.mark.xfail(reason="Does not warn on special codepath")
     def test_boost_eval_issue_14606(self):
         q, a, b = 0.995, 1.0e11, 1.0e13
         with pytest.warns(RuntimeWarning):
```

</details>

I was able to confirm failures with `invalid value encountered in _beta_ppf` before the patch and fix after using `clang-16` locally on ARM Mac.

tylerjereddy commented:
As for the second blocker in gh-20576, the `assert_no_overwrite` seems to be probing data overwrite/copy semantics to some degree, and if we diff against `main` for the failing custom class:

```diff
--- a/scipy/linalg/_testutils.py
+++ b/scipy/linalg/_testutils.py
@@ -12,8 +12,6 @@ class _FakeMatrix2:
         self._data = data
 
     def __array__(self, dtype=None, copy=None):
-        if copy:
-            return self._data.copy()
         return self._data
```

that seems like a pretty likely reason to deviate from `main` on behavior, so perhaps worth just pushing that in and seeing if it makes the Windows failure go away.

dschult committed:
BUG: sparse: Fix summing duplicates for CSR/CSC creation from (data,coords) (#20687)

* test and then fix duplicates for CSR/CSC creation from (data,coords)

* remove has_canonical_format check when summing duplicates.

dschmitz89 commented:
> @steppi @czgdp1807 @dschmitz89 what do you think about applying the patch below (similar to the one suggested at: [#20208 (comment)](https://github.com/scipy/scipy/issues/20208#issuecomment-2027547367)) to resolve the first part of #20576 on the release branch here? There's an explanation there about why I'm hesitant to backport [gh-20393](https://github.com/scipy/scipy/pull/20393) -- do you agree with my hesitation?
> 
> ```diff
> --- a/scipy/stats/_continuous_distns.py
> +++ b/scipy/stats/_continuous_distns.py
> @@ -692,12 +692,10 @@ class beta_gen(rv_continuous):
>          return _boost._beta_sf(x, a, b)
>  
>      def _isf(self, x, a, b):
> -        with np.errstate(over='ignore'):  # see gh-17432
> -            return _boost._beta_isf(x, a, b)
> +         return sc.betaincinv(a, b, 1 - x)
>  
>      def _ppf(self, q, a, b):
> -        with np.errstate(over='ignore'):  # see gh-17432
> -            return _boost._beta_ppf(q, a, b)
> +        return sc.betaincinv(a, b, q)
>  
>      def _stats(self, a, b):
>          return (
> diff --git a/scipy/stats/tests/test_distributions.py b/scipy/stats/tests/test_distributions.py
> index 6d91bb8a6..de26dc3d8 100644
> --- a/scipy/stats/tests/test_distributions.py
> +++ b/scipy/stats/tests/test_distributions.py
> @@ -4454,7 +4454,7 @@ class TestBeta:
>          assert_equal(stats.beta.pdf(1, a, b), 5)
>          assert_equal(stats.beta.pdf(1-1e-310, a, b), 5)
>  
> -    @pytest.mark.xfail(IS_PYPY, reason="Does not convert boost warning")
> +    @pytest.mark.xfail(reason="Does not warn on special codepath")
>      def test_boost_eval_issue_14606(self):
>          q, a, b = 0.995, 1.0e11, 1.0e13
>          with pytest.warns(RuntimeWarning):
> ```
> 
> I was able to confirm failures with `invalid value encountered in _beta_ppf` before the patch and fix after using `clang-16` locally on ARM Mac.

For the inverse survival function `_isf` it must be `return sc.betainccinv(a, b, x)` instead, besides that the patch looks fine. Agree that sorting out everything that happened in `special` would be a big pain to backport, so let's keep it simple like this.

tylerjereddy committed:
MAINT: release branch failures

* attempt to deal with gh-20576

tylerjereddy committed:
MAINT: import fixes

* some minor import fixes following a large
series of backports

tylerjereddy committed:
Revert "MAINT: stats: update boost to improve `skewnorm.ppf` (#20643)"

This reverts commit ee6203de3b77b76dc21be802b468fc2f7320f847.

tylerjereddy committed:
MAINT: import fixes

* more import cleanups after backport activity
to make the linter happy

tylerjereddy commented:
I've added substantially more backports and a few manual shims--we're back at parity with merged backport-labeled PRs.

Let's where the regular CI lands. I did have to revert and unlabel one PR (gh-20643) based on local testing with clang-16 on Mac, but the rest made it in, so far. I expanded the TODO list above based on known CI failure in `main`, etc.

tylerjereddy committed:
CI, MAINT: pin Python for MacOS conda

* We're seeing CI failures related to an undesirable
bump to Python `3.12` in this job, when the intention
was clearly to respect the Python version specific
in the GHA matrix. I didn't check too closely why
exactly it suddenly started happening, but some
packages weren't ready for `3.12` yet on this
job (`scikit-umfpack` in particular) and I don't
see too much harm in adding an extra pin to
respect the intention for the Python version.

mdhaber committed:
MAINT: stats.wilcoxon: fix bug with Ndim>1, shape[axis]>50, NaN, 'auto' (#20592)

tylerjereddy committed:
CI: MR 20632 revisions

* attempt to fix `M1 test - openblas` MacOS ARM CI job
based on some `gfortran` shims applied to a similar
job on `main`

tylerjereddy commented:
The regular CI is finally passing here (1 or 2 known flakes sometimes require restarts).

I'm going to proceed with a test pass of the wheel builds before updating the release notes again, because we may see shenanigans with the bump to OpenBLAS `0.3.27` for example.

tylerjereddy committed:
CI: PR 20632 revisions [wheel build]

* This is an empty commit to test the wheel
building in the above PR.

[wheel build]

tylerjereddy commented:
For MacOS ARM wheel failure `line 13: gfortran: command not found`, I may need similar shims to the regular CI Mac ARM job changes I made (based on a subset of changes from @thalassemia Acc support work I think), without backporting the full accelerate support stuff of course.

dschmitz89 committed:
BUG: stats: Fix `zipf.pmf` and `zipfian.pmf` for int32 `k` (#20702) [wheel build]

* BUG: fix zipf.pmf for integer k

* ENH: increase range for zipfian.pmf for integer k

---------

Co-authored-by: Tyler Reddy <tyler.je.reddy@gmail.com>

[wheel build]

tylerjereddy committed:
MAINT: PR 20632 revisions [wheel build]

* Changes needed to get MacOS ARM wheels building
again, based on debug work on my fork.

[wheel build]

tylerjereddy commented:
I pushed in a patch that helped deal with the MacOS ARM wheel build issue on my fork, but now of course there's another problem with the new `pythran` released yesterday. Hopefully I can patch that with less friction.

tylerjereddy committed:
MAINT: PR 20632 revisions [wheel build]

* Need to pin `pythran` version in a CI job
because of a recent new release.

[wheel build]

tylerjereddy committed:
CI, MAINT: PR 20632 revisions

* Increase the Cirrus CI Linux aarch64
job timeout to 72 minutes; it timed
out twice in a row via manual restarts.

tylerjereddy commented on .github/workflows/wheels.yml:
```diff
@@ -152,8 +147,8 @@ jobs:
-            CIBW="DYLD_LIBRARY_PATH=$GFORTRAN_LIB:/opt/arm64-builds/lib delocate-listdeps {wheel} &&\
-              DYLD_LIBRARY_PATH=$GFORTRAN_LIB:/opt/arm64-builds/lib delocate-wheel --require-archs {delocate_archs} -w {dest_dir} {wheel}"
+            CIBW="DYLD_LIBRARY_PATH=$GFORTRAN_LIB:$LIB_PATH delocate-listdeps {wheel} &&\
+              DYLD_LIBRARY_PATH=$GFORTRAN_LIB:$LIB_PATH delocate-wheel --require-archs {delocate_archs} -w {dest_dir} {wheel}"
```
@rgommers @h-vetinari This PR is taking shape I think--the full CI matrix (including wheels) is finally passing. The changes inline above are probably the most annoying/controversial ones I had to make manually (they are probably a superset of what is really needed to get Mac ARM wheels passing).

Any remaining objections/concerns for `1.13.1` release? We're at parity with the current list of merged backport-labeled PRs (30 in total now). I have two checklist items left above, but they are basically just updating the relnotes and reading the diff carefully before proceeding.

@mdhaber @dschult @perimosocordiae any more backports that are important for `stats` or `sparse` for `1.13.1`?

I think if folks are happy I'd try to get `1.13.1` out before branching `1.14.0`, with a modest delay likely for the latter, but hopefully not too substantial (I'm blocking off time for release notes work tomorrow).

rgommers replied:
> Any remaining objections/concerns for `1.13.1` release? We're at parity with the current list of merged backport-labeled PRs (30 in total now).

None that I can think of. This is already quite large, so +1 for shipping it soon.

The changes for macOS gfortran should be okay - whatever works there I think is fine in a maintenance branch.

mdhaber replied:
I didn't get to see this earlier, but I think this has all the backports I had in mind.

tylerjereddy commented on ci/cirrus_wheels.yml:
```diff
@@ -19,6 +19,7 @@ cirrus_wheels_linux_aarch64_task:
     platform: linux
     cpu: 2
     memory: 4G
+    timeout_in: 72m
```
I tried restarting twice; this is probably cheaper than needing to hit restart several times during the rel process.

rgommers replied:
+1 that makes sense to me, it was always close to the limit.

tylerjereddy committed:
DOC: PR 20632 updates

* Update the SciPy `1.13.1` release notes following substantial
backport activity.

[skip cirrus]

tylerjereddy commented:
A new test failure showed up in a Windows CI job (https://github.com/scipy/scipy/actions/runs/9194423643/job/25287826002?pr=20632).

<details>

```
_________________________ TestPoisson.test_fill_space _________________________

self = <scipy.stats.tests.test_qmc.TestPoisson object at 0x000002389148B210>

    def test_fill_space(self):
        radius = 0.2
        engine = self.qmce(d=2, radius=radius)
    
        sample = engine.fill_space()
        # circle packing problem is np complex
>       assert l2_norm(sample) >= radius

C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\scipy\stats\tests\test_qmc.py:938: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\scipy\stats\tests\test_qmc.py:1410: in l2_norm
    return distance.pdist(sample).min()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

a = array([], dtype=float64), axis = None, out = None, keepdims = False
initial = <no value>, where = True

    def _amin(a, axis=None, out=None, keepdims=False,
              initial=_NoValue, where=True):
>       return umr_minimum(a, axis, None, out, keepdims, initial, where)
E       ValueError: zero-size array to reduction operation minimum which has no identity

C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\numpy\_core\_methods.py:48: ValueError
============================== warnings summary ===============================
stats/tests/test_qmc.py::TestUtils::test_geometric_discrepancy_mst_with_zero_distances
  C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\scipy\stats\tests\test_qmc.py:232: UserWarning: Sample contains duplicate points.
    assert_allclose(qmc.geometric_discrepancy(sample, method='mst'), 0.5)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ===========================
FAILED stats/tests/test_qmc.py::TestPoisson::test_fill_space - ValueError: zero-size array to reduction operation minimum which has no identity
```

</details>

I'm going to restart that now and hope it is a fluke (cc @tupui since it is `qmc`, but I don't see any reports of it flaking in our issue tracker).

Beyond that, I read through the full diff again and didn't see anything I didn't expect. The rendered versions of the `1.13.1` and `1.13.0` (also modified here) release notes don't seem to have any major issues.

tylerjereddy commented:
Regular CI is green again, proceeding with `1.13.1` release process.

tylerjereddy commented:
We have a 1.5:1 commit:backport PR ratio, which is a bit high, but I'd rather not play with commit history/fire.

tylerjereddy commented:
somehow a Cirrus job still timed out at an hour or so during the rel process.. restarting it is I guess.
